Add support for Stripe webhooks

waiting-for-dev opened the pull request:
## Summary

We set up the backbone to support [Stripe webhook](https://stripe.com/docs/webhooks) events. There're two different use cases:

- Users want to listen to certain webhook events. They'll need to add their names in a `webhook_events` setting.
- This extension code makes internal use of some of them. That's something programmed, and when the time comes, we'll need to add the event names in the `SolidusStripe::Webhook::Event::CORE_EVENTS` constant.

A new `POST /webhooks` endpoint exists to support the integration. Users must also configure a `solidus_stripe.webhook_endpoint_secret` Rails credential to [validate the incoming events](https://stripe.com/docs/webhooks/signatures). There's also a `webhook_signature_tolerance` setting to change the default number of seconds [while a signature is valid](https://stripe.com/docs/webhooks/signatures#replay-attacks).

When an incoming webhook arrives, a matching `Spree::Bus` event prepended with `stripe.` is published. Therefore, all the usual techniques in [Solidus's pub/sub system](https://guides.solidus.io/customization/subscribing-to-events) are allowed.

I'd like to move the decoding of the event from a controller to a service, but as we don't have the infrastructure ready on Solidus is better to wait and not implement a custom solution here.

## Checklist

Check out our [PR guidelines](https://github.com/solidusio/.github/blob/master/CONTRIBUTING.md#pull-request-guidelines) for more details.

The following are mandatory for all PRs:

- [x] I have written a thorough PR description.
- [x] I have kept my commits small and atomic.
- [x] I have used clear, explanatory commit messages.

The following are not always needed (~cross them out~ if they are not):

- [x] I have added automated tests to cover my changes.
- ~[ ] I have attached screenshots to demo visual changes.~
- ~[ ] I have opened a PR to update the [guides](https://github.com/solidusio/edgeguides).~
- [x] I have updated the README to account for my changes.

kennyadsl commented on lib/solidus_stripe/configuration.rb:
```diff
@@ -1,10 +1,24 @@
+    #  @return [Array<Symbol>] stripe events to handle. You also need to
+    #  register them in the Stripe dashboard. For an event `:foo`, a matching
+    #  `:"stripe.foo"` event will be registered in {Spree::Bus}.
+    attr_accessor :webhook_events
```
Can user remove events that we need for the extension with this preference?

waiting-for-dev replied:
No, they can't, as we add them into the `SolidusStripe::Webhook::Event::CORE_EVENTS` constant.

kennyadsl commented on app/controllers/solidus_stripe/webhooks_controller.rb:
```diff
@@ -0,0 +1,39 @@
+    respond_to :json
+
+    def create
+      request.body.read
```
Thoughts on moving this to a service? I read your point that we are not ready with the infrastructure, and that makes sense. But when we will, and want to convert this, it will probably be easier because we can deprecate that class. 

No strong preference here: in the @elia's PR there are several controllers not using services, so this is more a general discussion about the approach we want to take.

elia replied:
Some of this could probably be kept inside `SolidusStripe::Webhook::Event`, something like:

```rb
def (SolidusStripe::Webhook::Event).from_request(request)
  tolerance = SolidusStripe.configuration.webhook_signature_tolerance
  secret = Rails.application.credentials.solidus_stripe[:webhook_endpoint_secret] # move to a configuration?
  signature_header = request.headers[SIGNATURE_HEADER]

  Stripe::Webhook.construct_event(payload, signature_header, secret, tolerance: tolerance)
end
```

waiting-for-dev replied:
Moving things to services requires some decisions to be made, like the interface we create for a result object. I think it doesn't pay off to build a skeleton here that will need to be deprecated later. At most, as @elia suggested, we can try to move some things into collaborators. The `from_request` method makes sense, but I'd also move the creation to the `SolidusStripe::Webhook::Event` there. This way, an instance of the collaborator will be involved. I'd mark the method as `@api private` so we're free to remove it later. Thoughts?

elia replied:
I think I need to see the code in order to understand fully the proposal, but I'm down with it from what I understand… and it's way easier to go from `@api private` to `@api public` than the other way around 🙌 

The main goal is to have building block that are easy to reuse in case the target app wants (or needs) to customize or redo the controller.

waiting-for-dev replied:
Please, take a look now. I moved the logic other than publishing the event to a class method in `SolidusStripe::Webhook::Event`.

kennyadsl commented on config/routes.rb:
```diff
@@ -2,4 +2,5 @@
 
 SolidusStripe::Engine.routes.draw do
   resources :payment_intents, only: :create
+  resources :webhooks, only: :create, format: false
```
I think there could be conflicts here if we don't use some prefixes. `/webhooks` has high chances to be already taken.

waiting-for-dev replied:
The generator will mount the routes within a `solidus_stripe` namespace, so I'll end up being `/solidus_stripe/webhooks`. Do you think that's enough? Or do you think people might end up mounting the engine on the root namespace?

kennyadsl replied:
That's perfect. I was confused reading the PR description when you mention `POST /webhooks`. Solved!

kennyadsl reviewed (commented):
Left some comments but overall looks great!

elia commented on app/controllers/solidus_stripe/webhooks_controller.rb:
```diff
@@ -0,0 +1,39 @@
+      request.body.read
+        .then { |payload| Stripe::Webhook.construct_event(payload, signature_header, secret, tolerance: tolerance) }
+        .then { |stripe_event| SolidusStripe::Webhook::Event.new(stripe_event: stripe_event) }
+        .then { |event| Spree::Bus.publish(event) }
```
What happens when we receive an unknown event? 
Is that ignored by the bus or raises an exception?
Should we only publish registered events?

(please excuse my ignorance of the new bus on this topic)

waiting-for-dev replied:
When an unknown event is received, an exception is raised, rendering the names of all registered events. That helps users be protected against typos. For instance, say that you subscribe by accident to an event `:"stripe.charge.succeded"` (there's a missing `e`). If we don't raise when there's an incoming `chage.succeeded` stripe event, users can have a hard time debugging.

More generally, getting an event you weren't anticipating is something unexpected, so I think an exception makes sense here.

What do you think?

elia replied:
I see the point, which brings me to another question. I thought that stripe sends out all events it can, or do we have control in the dashboard over what events we want to receive?

I'd say if we have control then it's fine to raise; otherwise we risk having a bunch of exceptions from all the tens of events sent by stripe: in that case maybe we can filter the list of what we want to forward.

waiting-for-dev replied:
My understanding is that stripe requires you to [register the events you're interested in on the dashboard](https://stripe.com/docs/webhooks/go-live#configure-webhook-settings).

waiting-for-dev replied:
A possible pain point would be for users listening to all events:
<img src="https://user-images.githubusercontent.com/52650/215456664-56ae1f32-409c-4c08-857f-cbf8670b5875.png" width="50%" />

However, Omnes is designed with safety in mind, so it doesn't allow publishing events if they haven't been registered before. We could make some changes upstream to introduce namespace events and different strategies for checking the event name. For instance, allowing any event starting with `stripe.` if the user prefers.

I'd say we should move forward now and work on it in a later stage if we think or find out that's a valid concern. Thoughts?

elia replied:
The fact that Stripe lets you list the events you wanna receive settles it for me, we just need to instruct to list only the ones supported out-of-the-box when installing (or provide the code needed to do it from the API) and that's it 🙌 

Thanks for explaining 👍

elia reviewed (commented):
A few questions/sugestions

waiting-for-dev commented on lib/solidus_stripe/engine.rb:
```diff
@@ -14,6 +14,16 @@ class Engine < Rails::Engine
 
+    initializer "solidus_stripe.pub_sub", after: "spree.core.pub_sub" do |app|
+      app.reloader.to_prepare do
+        require "solidus_stripe/webhook/event"
```
Good catch. Fixed.

waiting-for-dev commented on lib/solidus_stripe/webhook/event.rb:
```diff
@@ -0,0 +1,54 @@
+
+      # @api private
+      def self.register(user_events:, bus:, core_events: CORE_EVENTS)
+        (core_events + user_events).each do |event|
+          bus.register(:"stripe.#{event}")
+        end
```
If everything is going to be handed to the user application, then we can remove the distinction. However, if we provide default implementations, that's not the case, right? 🤔 Maybe I'm missing something, but we probably need to capture incoming webhooks for payment capture so that our models are updated accordingly. That's not something I imagine belonging to the copied templates but to the core models. Am I wrong?

elia replied:
If possible I would err on the side of allowing the target app to take control of core events. Stripe has many features and maybe an additional fraud check or external system integration needs to be added.

The implementations we provide out of the box should be more examples and (I think) allow breaking them apart leaving the glue code in the target application. That to me would be the best of both worlds.

I'm assuming it's possible to do that, but if not maybe the current division is already the best we can do

waiting-for-dev replied:
As you can create more than one subscriber for a given event, we're good at letting users add new stuff.

As per our subscribers, we should design them with the SRP in mind. For instance, if we need to do two things on `payment.succeeded`, let's create two subscribers for the same event and not a single one. This way, we pinpoint the need for any of the regular Solidus approaches to modify core behavior. On top of that, for subscribers, Omnes allows you to [unsubscribe](https://github.com/nebulab/omnes#unsubscribing--clearing) listeners from the bus 👍

waiting-for-dev commented:
I'm currently investigating the issue on CI on Ruby 2.7.

When doing:

```ruby
        .then { |stripe_event| SolidusStripe::Webhook::Event.new(stripe_event: stripe_event) }
        .then { |event| Spree::Bus.publish(event) }
```

The event yield on the second step is a legit `SolidusStripe::Webhook::Event`. However, `Spree::Bus#publish` receives a `Hash`. For now, I've narrowed down the issue to any object responding to `#to_hash`. `Stripe::Event` implements it, and `SolidusStripe::Webhook::Event` delegates to it.

elia commented:
@waiting-for-dev would probably be wiser to use `#to_h` which is the generic conversion form, `#to_hash` `#to_str` `#to_ary` usually are for objects that are intrinsically of the type you're converting to. E.g. a symbol has `#to_s` and `#to_sym` but not `#to_str`.

waiting-for-dev commented:
> @waiting-for-dev would probably be wiser to use `#to_h` which is the generic conversion form, `#to_hash` `#to_str` `#to_ary` usually are for objects that are intrinsically of the type you're converting to. E.g. a symbol has `#to_s` and `#to_sym` but not `#to_str`.

@elia, the `#to_hash` method is coming from the `Stripe::Event` object, which we don't control. `SolidusStripe::Webhook::Event` is simply delegating to it

elia commented:
> @elia, the #to_hash method is coming from the Stripe::Event object, which we don't control. SolidusStripe::Webhook::Event is simply delegating to it

Yes, I meant that's the root cause in my opinion, combined with 2.7 handling of keyword-arguments splatting and potentially the `then` implementation.

waiting-for-dev commented:
> > @elia, the #to_hash method is coming from the Stripe::Event object, which we don't control. SolidusStripe::Webhook::Event is simply delegating to it
> 
> Yes, I meant that's the root cause in my opinion, combined with 2.7 handling of keyword-arguments splatting and potentially the `then` implementation.

Thanks, @elia, it turned out to be related to an override in Solidus itself. See https://github.com/solidusio/solidus/pull/4875

waiting-for-dev commented:
@elia, @kennyadsl, I have addressed your suggestions. Please, could you take another look? I also have added the new settings to the initializer template (I had forgotten) and did some refactoring to the test fixture to easily share them.

kennyadsl reviewed (approved):
👍 Thanks @waiting-for-dev

elia reviewed (approved):
Awesome!

I tried to understand what's going on with the sqlite failure in the CI but couldn't from the diff, any ideas @waiting-for-dev?

waiting-for-dev commented:
> Awesome!
> 
> I tried to understand what's going on with the sqlite failure in the CI but couldn't from the diff, any ideas @waiting-for-dev?

We already talked about it offline, but for context this is the issue I talked in https://github.com/solidusio/solidus_stripe/pull/157#issuecomment-1403512955
